### ISSUE TITLE
Error: Cannot read property 'forEach' of undefined

### ISSUE THREAD

**Issue by osartun** [contentful/contentful-migration#249]
EDIT: I debugged this issue, see [my comment below](https://github.com/contentful/contentful-migration/issues/249#issuecomment-593464680).

---

I've read up on how to use Contentful migrations and wanted to try it out for the very first time. However, when I was running my migration script I got the error `Cannot read property 'forEach' of undefined`:

```
🚨  Migration unsuccessful
Unexpected token C in JSON at position 0

🚨  Error: Cannot read property 'forEach' of undefined

  TypeError: Cannot read property 'forEach' of undefined
      at run ([…]/node_modules/contentful-migration/built/bin/cli.js:166:24)
      at processTicksAndRejections (internal/process/task_queues.js:94:5)
```

The reason why I'm creating an issue here is because I believe that this error is in fact not coming from my migration script. According to the stack trace the error is coming from here:
https://github.com/contentful/contentful-migration/blob/be850708122f9cdd5a132d75711fa278bd0ca543/src/bin/cli.ts#L187

This issue seems to be related to https://github.com/contentful/contentful-migration/issues/183. However, #183 was solved by using a different technology (the python content management SDK) to accomplish the goal. I'd rather want this tool here to work though.

## Expected Behavior

Either a successful migration or an error message telling what exactly was wrong with my migration script.

## Actual Behavior

Inactionable error message / error within the tool itself.

<img width="755" alt="Screenshot 2020-02-25 at 17 37 56" src="https://user-images.githubusercontent.com/1381730/75269038-95813400-57f8-11ea-83cc-2957d3f84e39.png">

## Possible Solution

Either this is a bug in the contentful-migration CLI that needs to be fixed or the error messages should become more helpful / actionable.

## Steps to Reproduce

I don't want to publicly post our space ID, but also don't want to go through the effort of creating a separate repo for showcasing the error.
So, I'll just post the migration script in case this is of any help:

```js
// migrations/01-lowercase-slug.js

module.exports = function(migration) {
  const transformEntryForLocale = fromFields => {
    const prevSlug = fromFields.slug
    const newSlug =
      typeof prevSlug === 'string' ? prevSlug.toLowerCase() : prevSlug
    return { slug: newSlug }
  }

  migration.transformEntries({
    contentType: 'crmPage',
    from: ['slug'],
    to: ['slug'],
    shouldPublish: 'preserve',
    transformEntryForLocale,
  })

  migration.transformEntries({
    contentType: 'pcmPage',
    from: ['slug'],
    to: ['slug'],
    shouldPublish: 'preserve',
    transformEntryForLocale,
  })
}
```

## Environment

-   **Node Version**: `v12.14.1`
-   **Package Manager Version**: `yarn 1.21.1`
-   **Operating System**: Darwin Olivers-MBP.fritz.box 19.2.0 Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64 x86_64
-   **Package Version**: 1.2.13 (`contentful -v`)

**Comment by osartun** [comment]
Update: So, I'm currently looking into this issue with a debugger and I can confirm that, yes, this is a problem in `contentful-migration` itself.

The issue is that when a request fails the error handler assumes that the error message is a JSON string and parses it accordingly:
https://github.com/contentful/contentful-migration/blob/be850708122f9cdd5a132d75711fa278bd0ca543/src/bin/cli.ts#L143-L152

This might be true for errors coming from Contentful itself. However, the error this is caught at this point is coming out of Axios which is used inside the Contentful client.
Therefore this error message doesn't have the expected pattern and isn't a JSON string. Instead the error message is:
```
TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'Object'
    --- property 'en-US' closes the circle
```

Now, in this line we're trying to run `JSON.parse` on this string.
https://github.com/contentful/contentful-migration/blob/be850708122f9cdd5a132d75711fa278bd0ca543/src/bin/cli.ts#L143

Which fails, of course, hence the error message
```
Unexpected token C in JSON at position 0
```
Because at position 0 you have a `C` (because the error `message` is: `"Converting circular structure to JSON"`).

Now, because the script fails at this unexpected position it doesn't even reach the point where all the errors are put into one `BatchError`
https://github.com/contentful/contentful-migration/blob/be850708122f9cdd5a132d75711fa278bd0ca543/src/bin/cli.ts#L155-L158

Which is why there is no `errors` property at the `error` object received in this event handler:
https://github.com/contentful/contentful-migration/blob/be850708122f9cdd5a132d75711fa278bd0ca543/src/bin/cli.ts#L184-L187

Which is why you get the error message
```
  TypeError: Cannot read property 'forEach' of undefined
      at run ([…]/node_modules/contentful-migration/built/bin/cli.js:166:24)
      at processTicksAndRejections (internal/process/task_queues.js:94:5)
```
Because `err.errors` is `undefined`.

So, overall we have multiple levels of errors here:

1. There's something wrong with the config object passed to Axios which is why axios errors out
2. The `makeRequest` error handler assumes to only get Contentful server errors and isn't made to optimize other errors
3. The "migration unsuccessful" error handler assumes to only get the `BatchError` object and isn't made to handle other errors

**Comment by osartun** [comment]
So, I figured out how to resolve my specific problem. I didn't implement the `transformEntryForLocale` function correctly. I assumed that `fromFields` would look like this:
```js
// fromFields: My assumption
{
  slug: 'the-slug'
}
```
Instead it looks like this:
```js
// fromFields: The actual structure
{
  slug: {
    'de': 'the-de-slug',
    'en-US': 'the-en-us-slug',
    'fr': undefined
    // etc…
  }
}
```
Meaning that the iterator function needs to look like this:
```js
const transformEntryForLocale = (fromFields, locale) => {
  const prevSlug = fromFields.slug[locale]
  if (!prevSlug) {
    return
  }
  // The objective of this migration is to ensure that all slugs are lowercase
  const newSlug = prevSlug.toLowerCase()
  return { slug: newSlug }
}
```

-----

This means I was able to unblock myself with this. However, I don't want to close this issue because I laid out in the previous comment `contentful-migration` isn't able to deal with issues like this. By implementing this function incorrectly I essentially broke the tool. This is not what's supposed to happen.

**Comment by makinwab** [comment]
Hi Osartun, thanks for sharing your findings so far.  Just to let you know that we're currently looking into this.

**Comment by patrickdmccarthy** [comment]
@osartun How did you debug this? I'm having the same error, but there's a different root issue (I have the correct transformEntryForLocale config and have upgraded to the latest version.)  For me, it fails around the 80th request and not the first, which is strange.

**Comment by osartun** [comment]
@patrickdmccarthy So, my way of debugging this might not be the most elegant way: I went into my `node_modules` directory (!), searched for the `contentful-migration` folder and added some `console.log`s to it. Later, I also added `debugger` statements or breakpoints in VSCode and attached the Node debugger to the process (See here: https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_attach-to-node-process-action).
The code in my node_modules folder is the built code which obviously looks different than the source code in the git repo here, so I looked into the source code here and compared it to the code I read in my node_modules folder to see where I'm at. The source code is easier to read and to understand, of course.

I hope this helps you! 🙏
